mixer/pkg/runtime2/config/ephemeral.go

```diff
@@ -0,0 +1,383 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"istio.io/istio/mixer/pkg/adapter"
+	config "istio.io/istio/mixer/pkg/config/proto"
+	"istio.io/istio/mixer/pkg/config/store"
+	"istio.io/istio/mixer/pkg/expr"
+	"istio.io/istio/mixer/pkg/template"
+	"istio.io/istio/pkg/log"
+)
+
+// Ephemeral configuration state that gets updated by incoming config change events. By itself, the data contained
+// is not meaningful. BuildSnapshot must be called to create a new snapshot instance, which contains fully resolved
+// config.
+type Ephemeral struct {
+	// Static information
+	adapters  map[string]*adapter.Info
+	templates map[string]*template.Info
+
+	// next snapshot id
+	nextID int64
+
+	// whether the Attributes have changed since last snapshot
+	attributesChanged bool
+
+	// entries that are currently known.
+	entries map[store.Key]*store.Resource
+
+	// attributes from the last config state update. If the manifest hasn't changed since the last config update
+	// the attributes are reused.
+	cachedAttributes map[string]*config.AttributeManifest_AttributeInfo
+}
+
+// NewEphemeral returns a new Ephemeral instance.
+func NewEphemeral(
+	templates map[string]*template.Info,
+	adapters map[string]*adapter.Info) *Ephemeral {
+
+	e := &Ephemeral{
+		templates: templates,
+		adapters:  adapters,
+
+		nextID: 0,
+
+		attributesChanged: false,
+		entries:           make(map[store.Key]*store.Resource, 0),
+
+		cachedAttributes: nil,
+	}
+
+	// build the initial snapshot.
+	_ = e.BuildSnapshot()
+
+	return e
+}
+
+// SetState with the supplied state map. All existing ephemeral state is overwritten.
+func (e *Ephemeral) SetState(state map[store.Key]*store.Resource) {
+	e.entries = state
+
+	for k := range state {
+		if k.Kind == AttributeManifestKind {
+			e.attributesChanged = true
+			break
+		}
+	}
+}
+
+// ApplyEvents to the internal ephemeral state. This gets called by an external event listener to relay store change
+// events to this ephemeral config object.
+func (e *Ephemeral) ApplyEvents(events []*store.Event) {
+
+	if log.DebugEnabled() {
+		log.Debugf("Incoming config change events: count='%d'", len(events))
+	}
+
+	for _, event := range events {
+		e.ApplyEvent(event)
+	}
+}
+
+// ApplyEvent to the internal ephemeral state. This gets called by an external event listener to relay store change
+// events to this ephemeral config object.
+func (e *Ephemeral) ApplyEvent(event *store.Event) {
```

Comment:
should be a private function.

Reply:
Actually, I think I don't need ApplyEvents anymore. Removing it and updating tests to use the singular one.